src/spline2/BsplineAllocator.hpp

```diff
@@ -0,0 +1,263 @@
+        Ny = My+3;
+      else
+        Ny = My+2;
+      y_grid.delta = (y_grid.end - y_grid.start)/(double)(Ny-3);
+      y_grid.delta_inv = 1.0/y_grid.delta;
+      spline->y_grid   = y_grid;
+
+      if (zBC.lCode == PERIODIC || zBC.lCode == ANTIPERIODIC)
+        Nz = Mz+3;
+      else
+        Nz = Mz+2;
+      z_grid.delta = (z_grid.end - z_grid.start)/(double)(Nz-3);
+      z_grid.delta_inv = 1.0/z_grid.delta;
+      spline->z_grid   = z_grid;
+
+      const int N = getAlignedSize<real_type, ALIGN>(num_splines);
+
+      spline->x_stride = (size_t)Ny*(size_t)Nz*(size_t)N;
+      spline->y_stride = Nz*N;
+      spline->z_stride = N;
+
+      spline->coefs_size=(size_t)Nx*spline->x_stride;
+      spline->coefs = mAllocator.allocate(spline->coefs_size);
+
+      return spline;
+    }
+
+  template<typename T, size_t ALIGN, typename ALLOC>
+    typename BsplineAllocator<T,ALIGN,ALLOC>::SingleSplineType*
+    BsplineAllocator<T,ALIGN,ALLOC>::allocateUBspline(Ugrid x_grid, Ugrid y_grid, Ugrid z_grid,
+        BCType xBC, BCType yBC, BCType zBC, T* data)
+    {
+      // Create new spline
+      SingleSplineType* restrict spline = new SingleSplineType;
+      spline->spcode = bspline_traits<T,3>::single_spcode;
+      spline->tcode  = bspline_traits<T,3>::tcode;
+      spline->xBC = xBC;
+      spline->yBC = yBC;
+      spline->zBC = zBC;
+
+      // Setup internal variables
+      int Mx = x_grid.num;  int My = y_grid.num; int Mz = z_grid.num;
+      int Nx, Ny, Nz;
+
+      if (xBC.lCode == PERIODIC || xBC.lCode == ANTIPERIODIC)
+        Nx = Mx+3;
+      else
+        Nx = Mx+2;
+      x_grid.delta = (x_grid.end - x_grid.start)/(double)(Nx-3);
+      x_grid.delta_inv = 1.0/x_grid.delta;
+      spline->x_grid   = x_grid;
+
+      if (yBC.lCode == PERIODIC || yBC.lCode == ANTIPERIODIC)
+        Ny = My+3;
+      else
+        Ny = My+2;
+      y_grid.delta = (y_grid.end - y_grid.start)/(double)(Ny-3);
+      y_grid.delta_inv = 1.0/y_grid.delta;
+      spline->y_grid   = y_grid;
+
+      if (zBC.lCode == PERIODIC || zBC.lCode == ANTIPERIODIC)
+        Nz = Mz+3;
+      else
+        Nz = Mz+2;
+      z_grid.delta = (z_grid.end - z_grid.start)/(double)(Nz-3);
+      z_grid.delta_inv = 1.0/z_grid.delta;
+      spline->z_grid   = z_grid;
+
+      spline->x_stride = Ny*Nz;
+      spline->y_stride = Nz;
+
+      spline->coefs_size=(size_t)Nx*(size_t)Ny*(size_t)Nz;
+
+      spline->coefs = mAllocator.allocate(spline->coefs_size);
+
+      if(data != NULL) // only data is provided
+      {
+        // First, solve in the X-direction
+        #pragma omp parallel for
+        for (int iy=0; iy<My; iy++)
+          for (int iz=0; iz<Mz; iz++) {
+            intptr_t doffset = iy*Mz+iz;
+            intptr_t coffset = iy*Nz+iz;
+            einspline::find_coefs_1d(spline->x_grid, xBC, data+doffset, My*Mz,
+                            spline->coefs+coffset, Ny*Nz);
+          }
+
+        // Now, solve in the Y-direction
+        #pragma omp parallel for
```

Comment:
These pragma omp parallel for's are the only thing that has come over from einspline that bothers me.  But I expanded the test program and they do help in the case of larger grids.  It seems for the smaller sizes we generally use that they don't do a great deal ~15%.

Reply:
Currently, the splines are constructed sequentially. Threads are anyway idle and gaining some thing is always good.